### ISSUE TITLE
$File::Find::name ne $_ with no_chdir

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#5033]
Migrated from [rt.perl.org#8522](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=8522) (status was 'resolved')

Searchable as RT8522$

**Comment by p5pRT** [comment]
### From anders@ieee.org

#### Created by anders@ieee.org

This is a bug report for perl from anders@&#8203;ieee\.org\,
generated with the help of perlbug 1\.28 running under perl v5\.6\.0\.


\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
Using File&#8203;::Find\, if find\(\) is called with \{no\_chdir => 1\}\, and the file/dir
argument is an ordinary file not in the current directory\, then
$File&#8203;::Find&#8203;::name will not be the same as $\_ within the wanted\(\)
subroutine\. This contradicts the documentation\.

The work\-around is to do this inside the wanted\(\) subroutine&#8203;:

 local\($\_\)=$File&#8203;::Find&#8203;::name;

Here is a patch \(relative to Perl 5\.7\.2\) containing both a test and a fix&#8203;:

\-\-\-\-\- Begin Patch \-\-\-\-\-

diff \-Naur perl\-5\.7\.2/lib/File/Find/find\.t
perl\-5\.7\.2\_find/lib/File/Find/find\.t
<details><summary>Inline Patch</summary>

```diff
--- perl-5.7.2/lib/File/Find/find.t Mon Jul  9 07:10:34 2001
+++ perl-5.7.2_find/lib/File/Find/find.t Tue Feb 12 23:31:25 2002
@@ -15,8 +15,8 @@
     $SIG{'__WARN__'} = sub { $warn_msg = $_[0]; warn "# $_[0]"; }
 }

-if ( $symlink_exists ) { print "1..188\n"; }
-else                   { print "1..78\n";  }
+if ( $symlink_exists ) { print "1..189\n"; }
+else                   { print "1..79\n";  }

 use File::Find;
 use File::Spec;
@@ -738,5 +738,10 @@
     Check( scalar(keys %Expect_File) == 0 );
     unlink file_path('fa', 'faa_sl');

+    my $name_eq_under=1;
+    File::Find::find( {wanted => sub {
+       undef $name_eq_under unless $File::Find::name eq $_;
+             }, no_chdir => 1}, file_path_name('fa', 'fa_ord') );
+    Check( $name_eq_under );
 }

diff -Naur perl-5.7.2/lib/File/Find.pm perl-5.7.2_find/lib/File/Find.pm
--- perl-5.7.2/lib/File/Find.pm Mon Jul  9 07:10:34 2001
+++ perl-5.7.2_find/lib/File/Find.pm Tue Feb 12 23:39:16 2002
@@ -602,6 +602,7 @@
      }

      $name = $abs_dir . $_; # $File::Find::name
+     $_ = $name if $no_chdir;

      { &$wanted_callback }; # protect against wild "next"

----- End Patch -----
```
</details>


<details><summary>Perl Info</summary>

```

Flags:
    category=library
    severity=low

Site configuration information for perl v5.6.0:

Configured by root at Fri May 11 15:15:27 GMT 2001.

Summary of my perl5 (revision 5.0 version 6 subversion 0) configuration:
  Platform:
    osname=linux, osvers=2.4.3, archname=i586-linux
    uname='linux subbotin 2.4.3 #1 tue may 8 21:54:34 gmt 2001 i686 unknown
'
    config_args='-ds -e -Dprefix=/usr -Di_db -Di_dbm -Di_ndbm -Di_gdbm'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef
usemultiplicity=undef
    useperlio=undef d_sfio=undef uselargefiles=define
    use64bitint=undef use64bitall=undef uselongdouble=undef usesocks=undef
  Compiler:
    cc='cc', optimize='-O2 -pipe', gccversion=2.95.3 20010315 (SuSE)
    cppflags='-fno-strict-aliasing -I/usr/local/include'
    ccflags
='-fno-strict-aliasing -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFS
ET_BITS=64'
    stdchar='char', d_stdstdio=define, usevfork=false
    intsize=4, longsize=4, ptrsize=4, doublesize=8
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t',
lseeksize=8
    alignbytes=4, usemymalloc=n, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -ldl -lm -lc -lcrypt
    libc=, so=so, useshrplib=false, libperl=libperl.a
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:



@INC for perl v5.6.0:
    /usr/lib/perl5/5.6.0/i586-linux
    /usr/lib/perl5/5.6.0
    /usr/lib/perl5/site_perl/5.6.0/i586-linux
    /usr/lib/perl5/site_perl/5.6.0
    /usr/lib/perl5/site_perl
    .


Environment for perl v5.6.0:
    HOME=/home/anders
    LANG=en_US
    LANGUAGE=english
    LC_COLLATE=POSIX
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)

PATH=/opt/kde2/bin:/home/anders/bin:/usr/local/bin:/usr/bin:/usr/X11R6/bin:/
bin:/usr/lib/java/bin:/usr/games/bin:/usr/games:/opt/gnome/bin:/opt/kde2/bin
:/opt/kde/bin:.:/opt/gnome/bin:/opt/gnome/bin::/opt/gnome/bin:
    PERL_BADLANG (unset)
    SHELL=/bin/tcsh




```
</details>

**Comment by p5pRT** [comment]
### From @smpeters

> \[anders \- Wed Feb 13 01&#8203;:18&#8203;:58 2002\]&#8203;:
> 
> This is a bug report for perl from anders@&#8203;ieee\.org\,
> generated with the help of perlbug 1\.28 running under perl v5\.6\.0\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> Using File&#8203;::Find\, if find\(\) is called with \{no\_chdir => 1\}\, and the file/dir
> argument is an ordinary file not in the current directory\, then
> $File&#8203;::Find&#8203;::name will not be the same as $\_ within the wanted\(\)
> subroutine\. This contradicts the documentation\.
> 

The documentation was changed prior to the creation of this patch to document the fact that 
$\_ will not necessarily be the same as $\_ when running with C\<chdir => 1>\.

**Comment by p5pRT** [comment]
@smpeters - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From @smpeters

> \[stmpeters \- Thu Jun 09 17&#8203;:49&#8203;:36 2005\]&#8203;:
> 
> > \[anders \- Wed Feb 13 01&#8203;:18&#8203;:58 2002\]&#8203;:
> >
> > This is a bug report for perl from anders@&#8203;ieee\.org\,
> > generated with the help of perlbug 1\.28 running under perl v5\.6\.0\.
> >
> >
> > \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> > Using File&#8203;::Find\, if find\(\) is called with \{no\_chdir => 1\}\, and the
> file/dir
> > argument is an ordinary file not in the current directory\, then
> > $File&#8203;::Find&#8203;::name will not be the same as $\_ within the wanted\(\)
> > subroutine\. This contradicts the documentation\.
> >
> 
> The documentation was changed prior to the creation of this patch to
> document the fact that
> $\_ will not necessarily be the same as $\_ when running with C\<chdir =>
> 1>\.
> 

Wait\, not in this case though\.  This problem was actually fix with a patch from ticket \#21031 
in change \#18809\.
